### PR TITLE
Guard against opaques crashing the compiler

### LINKED ISSUE #19434: Illegal MatchType Error reaches the backend
## Compiler version

3.4.0

## Minimized example

```Scala
object Test:

  object Named:
    opaque type Named[name <: String & Singleton, A] >: A = A

  type DropNames[T <: Tuple] = T match
    case Named.Named[_, x] *: xs => x *: DropNames[xs]
    case _ => T

  def f[T <: Tuple]: DropNames[T] = ???
```

## Output

```scala
an unexpected type representation reached the compiler backend while compiling named-tuples-strawman.scala: dotty.tools.dotc.core.Types$PreviousErrorType@2d9dff65. If possible, please file a bug on https://github.com/lampepfl/dotty/issues
scala.MatchError: dotty.tools.dotc.core.Types$PreviousErrorType@2d9dff65 (of class dotty.tools.dotc.core.Types$PreviousErrorType)
	at dotty.tools.backend.jvm.BCodeHelpers.dotty$tools$backend$jvm$BCodeHelpers$$typeToTypeKind(BCodeHelpers.scala:764)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind(BCodeHelpers.scala:202)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind$(BCodeHelpers.scala:130)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.toTypeKind(BCodeSkelBuilder.scala:134)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.asmMethodType(BCodeHelpers.scala:188)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.asmMethodType$(BCodeHelpers.scala:130)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.asmMethodType(BCodeSkelBuilder.scala:134)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:815)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:693)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen$$anonfun$1(BCodeSkelBuilder.scala:699)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:699)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:293)
	at dotty.tools.backend.jvm.CodeGen.genClass(CodeGen.scala:154)
	at dotty.tools.backend.jvm.CodeGen.genClassDef$1(CodeGen.scala:62)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:117)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1$$anonfun$1(CodeGen.scala:115)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:115)
	at dotty.tools.backend.jvm.CodeGen.genUnit(CodeGen.scala:120)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:83)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:354)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:360)
	at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:91)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:348)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:357)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:357)
	at dotty.tools.dotc.Run.compileSources(Run.scala:261)
	at dotty.tools.dotc.Run.compile(Run.scala:246)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
	at dotty.tools.dotc.Driver.process(Driver.scala:197)
	at dotty.tools.dotc.Driver.process(Driver.scala:165)
	at dotty.tools.dotc.Driver.process(Driver.scala:177)
	at dotty.tools.dotc.Driver.main(Driver.scala:207)
	at dotty.tools.dotc.Main.main(Main.scala)
Error while emitting named-tuples-strawman.scala
dotty.tools.dotc.core.Types$PreviousErrorType@2d9dff65 (of class dotty.tools.dotc.core.Types$PreviousErrorType)
1 warning found
1 error found
```

## Expectation

A regular error, like this one:
```
1 |object Test:
  |       ^
  |       The match type contains an illegal case:
  |           case Test.Named.Named[_, x] *: xs => x *: Test.DropNames[xs]
  |       (this error can be ignored for now with `-source:3.3`)
```
I verified that that was the error that generated the illegal error type.

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #19434

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -3570,7 +3570,8 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
       MatchResult.Stuck
 
     def recur(remaining: List[MatchTypeCaseSpec]): Type = remaining match
-      case (cas: MatchTypeCaseSpec.LegacyPatMat) :: _ if sourceVersion.isAtLeast(SourceVersion.`3.4`) =>
+      case (cas: MatchTypeCaseSpec.LegacyPatMat) :: _ if sourceVersion.isAtLeast(SourceVersion.`3.4`)
+          && (ctx.phaseId < ctx.base.elimOpaque.id || !ctx.base.elimOpaque.exists) =>
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3574]: Er ... I'm very uneasy about it. We should only do this as a very last resort. As is, we might (and probably will) reduce a match type differently before and after `elimOpaque`. IMO it's better to actually crash than to end up in that situation.

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3574]: At the very least, this should say *workaround* for #19434 with a big fat HACK/FIXME comment. It's not a real fix.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3574]: I don't know what the right fix is.  We have an applied match type alias and the compiler is asked to reduce it some time later.  But now it's a plain alias, so it's "illegal".  What should we do?  In this case we don't reduce, and I'm happy with that.  We just need to find where to tell it not to.

**Comment by odersky** [comment]
One possibility would be to merge ElimOpaque with erasure. It's a tiny phase. I am not sure anything between ElimOpaque and Erasure depends on the fact that opaque aliases are now aliases. So far no phase _declares_ that it needs to run after ElimOpaque but that's of course no guarantee that there are no dependencies.

**Comment by dwijnand** [comment]
I could see that fixing this case, seeing as the code branch is run in `preErasureCtx`.

**Comment by odersky** [comment]
Hmm. Maybe we have to leave ElimOpaque where it is since it seems to upset quite a few assumptions afterwards. We should document in the relevant phases thatthey have torun after ElimOpaque. 

That means that match type reduction has to cope with this somehow. One trick we could play is to refuse to normalize match types after ElimOpaque. Instead, time-travel to the ElimOpaque phase and do the reduction there. Or else treat normal aliases like opaque aliases, if the symbol has Opaque set before ElimOpaque. Not ure whether the second one is safe, though.

**Comment by dwijnand** [comment]
> Instead, time-travel to the ElimOpaque phase and do the reduction there. Or else treat normal aliases like opaque aliases, if the symbol has Opaque set before ElimOpaque. Not sure whether the second one is safe, though.

I didn't quite understand what you meant here.  Could you explain again what you think we should try?  Currently I'm just phase travelling to typer.

**Comment by sjrd** [comment]
My take on that idea is that we could try an `ElimMatchTypes` before `ElimOpaues`. It would replace every match type that reduces to its reduction, and for the others leave them as shells that only keep the upper bound and can never further reduce.

I'm not sure whether that's sound TBH.

**Comment by dwijnand** [comment]
Can't make that work.  It fails Ycheck.

**Pull request closed without merging**